crates/aptos-lint/src/lint/rules/unused_borrow_global_mut.rs

```diff
@@ -0,0 +1,137 @@
+use crate::lint::visitor::{ExpDataVisitor, LintUtilities};
+/// Detect all unused mutable variables, including those from borrow_global_mut, table::borrow_mut,
+/// vector::borrow_mut, etc.
+/// In these cases, immutable references should be obtained and used instead.
+use move_model::ast::{ExpData, Operation, Pattern};
+use move_model::{
+    model::{FunctionEnv, GlobalEnv, Loc, NodeId},
+    ty::{ReferenceKind, Type},
+};
+
+#[derive(Debug)]
+pub struct UnusedBorrowGlobalMutVisitor {
+    variables: Vec<VarInfo>,
+}
+
+#[derive(Debug, PartialEq, Clone, Eq, Hash)]
+struct VarInfo {
+    node_id: NodeId,
+    loc: Loc,
+    var_name: String,
+    used: bool,
+}
+impl VarInfo {
+    fn mark_used(&mut self) {
+        self.used = true;
+    }
+}
+
+impl Default for UnusedBorrowGlobalMutVisitor {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl UnusedBorrowGlobalMutVisitor {
+    pub fn new() -> Self {
+        Self {
+            variables: Vec::new(),
+        }
+    }
+
+    fn add_variable(&mut self, info: VarInfo) {
+        self.variables.push(info);
+    }
+
+    fn find_nearest_declaration(&mut self, name: String) {
+        for var in self.variables.iter_mut().rev() {
+            if var.var_name == name {
+                var.mark_used();
+                break;
+            }
+        }
+    }
+
+    pub fn visitor() -> Box<dyn ExpDataVisitor> {
+        Box::new(Self::new())
+    }
+
+    fn emit_warning(&mut self, env: &GlobalEnv) {
+        self.variables
+            .iter()
+            .filter(|x| !x.used)
+            .for_each(|var| {
+                let node_type = env.get_node_type(var.node_id);
+                if let Type::Reference(kind, _) = node_type {
+                    if ReferenceKind::Mutable == kind {
+                        let message = format!(
+                            "Unused borrowed mutable variable {}. Consider normal borrow (borrow_global, vector::borrow, etc.) instead",
+                            var.var_name
+                        );
+                        self.add_diagnostic_and_emit(
+                            &var.loc,
+                            &message,
+                            codespan_reporting::diagnostic::Severity::Warning,
+                            env
+                        );
+                    }
+                }
+            });
+    }
+
+    fn clear_variables(&mut self) {
+        self.variables.clear();
+    }
+
+    fn visit_exp_custom(&mut self, exp: &ExpData, env: &GlobalEnv) {
```

Comment:
In general, I think a lint like this may be better implemented on stackless bytecode (using a definition-use chain computed from dataflow analysis), because of several shortcomings in this implementation.